cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ArtifactHttpHandler.java

```diff
@@ -440,13 +443,56 @@ public void getArtifactPlugins(HttpRequest request, HttpResponder responder,
     } catch (IOException e) {
       LOG.error("Exception looking up plugins for artifact {}", artifactId, e);
       responder.sendString(HttpResponseStatus.INTERNAL_SERVER_ERROR,
-        "Error reading plugins for the artifact from the store.");
+                           "Error reading plugins for the artifact from the store.");
+    }
+  }
+
+  @Beta
+  @POST
+  @Path("/namespaces/{namespace-id}/artifacts/{artifact-name}/" +
+    "versions/{artifact-version}/plugintypes/{plugin-type}/plugins/{plugin-name}/methods/{plugin-method}")
+  public void callArtifactPluginMethod(HttpRequest request, HttpResponder responder,
+                                       @PathParam("namespace-id") String namespaceId,
+                                       @PathParam("artifact-name") String artifactName,
+                                       @PathParam("artifact-version") String artifactVersion,
+                                       @PathParam("plugin-name") String pluginName,
+                                       @PathParam("plugin-type") String pluginType,
+                                       @PathParam("plugin-method") String methodName,
+                                       @QueryParam("scope") @DefaultValue("user") String scope)
+    throws NamespaceNotFoundException, BadRequestException, ArtifactNotFoundException, IOException {
+
+    String requestBody = request.getContent().toString(Charsets.UTF_8);
+    Id.Namespace namespace = Id.Namespace.from(namespaceId);
+    Id.Namespace artifactNamespace = validateAndGetScopedNamespace(namespace, scope);
+    Id.Artifact artifactId = validateAndGetArtifactId(artifactNamespace, artifactName, artifactVersion);
+    if (requestBody.isEmpty()) {
+      responder.sendJson(HttpResponseStatus.BAD_REQUEST, "Request body is used as plugin method parameter, " +
+        "Received empty request body. ");
+      return;
+    }
+    try {
+      String response =
+        artifactRepository.pickParentArtifactAndInvokePluginEndpoint(namespace, artifactId, pluginType,
```

Comment:
What a method name. Why not just `invokePluginEndpoint`. Caller shouldn't care what's the implementation is (the pick parent artifact part).